File: Python/tests/test_chksum.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import pytest

from pathlib import Path
from hashlib import sha1
from base64 import urlsafe_b64encode as b64us

from seguid import slseguid
from seguid import scseguid
from seguid import dlseguid
from seguid import dcseguid

from seguid.chksum import seguid

from seguid.manip import reverse
from seguid.manip import rc
from seguid.manip import min_rotation_py
from seguid.manip import rotate_to_min

from seguid.reprutils import repr_from_tuple

from seguid.tables import COMPLEMENT_TABLE_DNA
from seguid.tables import TABLE_IUPAC_PROTEIN


def test_min_rotation():
    """Tests for the pydivsufsort min_rotation"""
    from pydivsufsort import min_rotation

    def smallest_rotation(s):
        i = min_rotation(s)
        return s[i:] + s[:i]

    assert rotate_to_min("taaa") == "aaat"
    assert (
        rotate_to_min("abaabaaabaababaaabaaababaab")
        == "aaabaaababaababaabaaabaabab"
    )
    assert (
        rotate_to_min("abaabaaabaababaaabaaaBabaab")
        == "Babaababaabaaabaababaaabaaa"
    )


def test_min_rotation_py():
    """docstring."""

    assert rotate_to_min("TAAA", min_rotation = min_rotation_py) == "AAAT"
    assert (
        rotate_to_min("ACAACAAACAACACAAACAAACACAAC",
                      min_rotation = min_rotation_py)
        == "AAACAAACACAACACAACAAACAACAC"
    )


def test_seguid():
    assert seguid("AT") == "seguid:Ax/RG6hzSrMEEWoCO1IWMGska+4"
    NP_313053_1 = (
        "MKALTARQQEVFDLIRDHISQTGMPPTRAEIAQRLGFRSPNAAEEHLKALARKGVIEIVSG"
        "ASRGIRLLQEEEEGLPLVGRVAAGEPLLAQQHIEGHYQVDPSLFKPNADFLLRVSGMSMKD"
        "IGIMDGDLLAVHKTQDVRNGQVVVARIDDEVTVKRLKKQGNKVELLPENSEFKPIVVDLRQ"
        "QSFTIEGLAVGVIRNGDWL"
    )

    assert seguid(NP_313053_1, table=TABLE_IUPAC_PROTEIN) == "seguid:2c4yjE+JqjvzYF1d0OmUh8pCpz8"




def cs(arg):
    return (
        b64us(sha1(arg.encode("ASCII")).digest()).decode("ASCII").rstrip("=")
    )


def test_slseguid():
    assert slseguid("AT") == "slseguid:Ax_RG6hzSrMEEWoCO1IWMGska-4"
    assert cs("AT") in slseguid("AT")


def test_scseguid():
    m13dna = Path("test_data/M13.txt").read_text().strip()
    sc = "scseguid:aAjgnsF9cPI6cu8IQ81sYnstVzU"
    assert scseguid(m13dna) == sc
    assert cs(Path("test_data/M13msg.txt").read_text().strip()) in sc


def test_dlseguid():
    ct = COMPLEMENT_TABLE_DNA
    table = ct | {"\n":"\n", "-":"-"}
    # AT
    # TA

    dlDNA = "AT"
    dlDNA_dlseguid = "AWD-dt5-TEua8RbOWfnctJIu9nA"
    assert dlseguid(dlDNA, rc(dlDNA), 0) == f"dlseguid:{dlDNA_dlseguid}"
    assert dlDNA_dlseguid in slseguid("AT\nTA", table = table)
    assert cs("AT\nTA") == dlDNA_dlseguid

    #  -AT
    #  AT-

    dlDNA2 = ("AT", "TA", 1)
    dlDNA2_dlseguid = "JwB2eUmZkCNjyWAv471JeUbiSDM"
    assert dlseguid(*dlDNA2) == f"dlseguid:{dlDNA2_dlseguid}"
    assert dlDNA2_dlseguid in slseguid("-AT\nAT-", table = table)
    assert cs("-AT\nAT-") == dlDNA2_dlseguid

    # TA-
    # -TA

    dlDNA3 = ("TA", "AT", -1)
    dlDNA3_dlseguid = "bv0UOR12eWrBeaAx79PNZvveviU"
    assert dlseguid(*dlDNA3) == f"dlseguid:{dlDNA3_dlseguid}"
    assert dlDNA3_dlseguid in slseguid("AT-\n-AT", table = table)
    assert cs("AT-\n-AT") == dlDNA3_dlseguid

    # CTATAG
    #   TA

    dlDNA4 = ("CTATAG", "AT", -2)
    dlDNA4_dlseguid = "np3hncfQvOh8rZ8Co1Ts_02NXg4"
    assert dlseguid(*dlDNA4) == f"dlseguid:{dlDNA4_dlseguid}"
    assert dlDNA4_dlseguid in slseguid("--AT--\nGATATC", table = table)
    assert cs("--AT--\nGATATC") == dlDNA4_dlseguid

    #   AT
    # GATATC

    dlDNA5 = ("AT", "CTATAG", 2)
    dlDNA5_dlseguid = "np3hncfQvOh8rZ8Co1Ts_02NXg4"
    assert dlseguid(*dlDNA5) == f"dlseguid:{dlDNA4_dlseguid}"
    assert dlDNA5_dlseguid in slseguid("--AT--\nGATATC", table = table)
    assert cs("--AT--\nGATATC") == dlDNA5_dlseguid

    repr_from_tuple("AT", "CTATAG", 2)


def test_dcseguid():
    pUC19dna = Path("test_data/puc19.txt").read_text().strip()
    dcsg = "dcseguid:zhw8Yrxfo3FO5DDccx4PamBVPCQ"
    assert dcseguid(pUC19dna, rc(pUC19dna)) == dcsg
    w, c = Path("test_data/pUC19msg.txt").read_text().splitlines()
    assert dlseguid(w, reverse(c), 0) == "dlseguid:zhw8Yrxfo3FO5DDccx4PamBVPCQ"

File: r/tests/test_chksum.R
library(seguid)

reverse <- seguid:::reverse

assert_error <- function(expr, envir = parent.frame()) {
  expr <- substitute(expr)
  res <- tryCatch(eval(expr, envir = envir), error = identity)
  if (!inherits(res, "error")) {
    stop("Call did not result in an error: ", deparse(expr))
  }
}


stopifnot(  seguid("AT") ==   "seguid=Ax/RG6hzSrMEEWoCO1IWMGska+4")
stopifnot(lsseguid("AT") == "lsseguid=Ax_RG6hzSrMEEWoCO1IWMGska-4")

NP_313053_1 <- paste0(
  "MKALTARQQEVFDLIRDHISQTGMPPTRAEIAQRLGFRSPNAAEEHLKALARKGVIEIVSG",
  "ASRGIRLLQEEEEGLPLVGRVAAGEPLLAQQHIEGHYQVDPSLFKPNADFLLRVSGMSMKD",
  "IGIMDGDLLAVHKTQDVRNGQVVVARIDDEVTVKRLKKQGNKVELLPENSEFKPIVVDLRQ",
  "QSFTIEGLAVGVIRNGDWL"
)
#stopifnot(seguid(NP_313053_1, alphabet="{IUPAC}") == "seguid=2c4yjE+JqjvzYF1d0OmUh8pCpz8")


m13dna <- readLines("test_data/M13.txt")
truth <- "csseguid=aAjgnsF9cPI6cu8IQ81sYnstVzU"
stopifnot(csseguid(m13dna) == truth)


truth <- "ldseguid=AWD-dt5-TEua8RbOWfnctJIu9nA"
stopifnot(ldseguid("AT", "AT") == truth)

truth <- "ldseguid=JwB2eUmZkCNjyWAv471JeUbiSDM"
stopifnot(ldseguid("-AT", "-TA") == truth)

truth <- "ldseguid=XBcVadfQevTW_lklW4rdqw5udQ8"
stopifnot(ldseguid("-TA", "-AT") == truth)

truth <- "ldseguid=_E05Xeo7KnLxrjsqDdpXNw_AIDE"
stopifnot(ldseguid("CTATAG", "--AT--") == truth)

truth <- "ldseguid=np3hncfQvOh8rZ8Co1Ts_02NXg4"
stopifnot(ldseguid("--AT--", "CTATAG") == truth)

truth <- "cdseguid=tYeHZYwxQGDHTqGDcrebERag0AU"
stopifnot(cdseguid("ACGTT", "AACGT") == truth)
stopifnot(cdseguid("AACGT", "ACGTT") == truth)

pUC19dna <- readLines("test_data/pUC19.txt", warn = FALSE)
pUC19dna_rc <- reverse(chartr("ACGT", "TGCA", pUC19dna))
truth <- "cdseguid=zhw8Yrxfo3FO5DDccx4PamBVPCQ"
stopifnot(cdseguid(pUC19dna, pUC19dna_rc) == truth)
bfr <- readLines("test_data/pUC19_minimal_rotation_watson_linebreak_crick.txt", warn = FALSE)
w <- bfr[1]
c <- bfr[2]
stopifnot(ldseguid(w, reverse(c)) == gsub("cdseguid=", "ldseguid=", truth))


## Empty input is considered an error
assert_error(seguid::seguid(""))
assert_error(seguid::lsseguid(""))
assert_error(seguid::csseguid(""))
assert_error(seguid::cdseguid("", ""))



## Use checksums as filenames
seq <- "GATTACA"
## Comment:
## The   SEGUID check is seguid=tp2jzeCM2e3W4yxtrrx09CMKa/8
## The slSEGUID check is seguid=tp2jzeCM2e3W4yxtrrx09CMKa_8
td <- tempdir()
filename <- seguid::lsseguid(seq)
pathname <- file.path(td, filename)
cat(seq, file = pathname)
stopifnot(utils::file_test("-f", pathname))
content <- readLines(pathname, warn = FALSE)
stopifnot(identical(content, seq))
file.remove(pathname)
unlink(td)
